File: tests/algorithms/x/PHP/physics/lorentz_transformation_four_vector.php
<?php
error_reporting(E_ALL & ~E_DEPRECATED);
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _panic($msg) {
    fwrite(STDERR, strval($msg));
    exit(1);
}
$__start_mem = memory_get_usage();
$__start = _now();
  $c = 299792458.0;
  function sqrtApprox($x) {
  global $c, $v;
  if ($x <= 0.0) {
  return 0.0;
}
  $guess = $x / 2.0;
  $i = 0;
  while ($i < 20) {
  $guess = ($guess + $x / $guess) / 2.0;
  $i = $i + 1;
};
  return $guess;
};
  function beta($velocity) {
  global $c, $v;
  if ($velocity > $c) {
  _panic('Speed must not exceed light speed 299,792,458 [m/s]!');
}
  if ($velocity < 1.0) {
  _panic('Speed must be greater than or equal to 1!');
}
  return $velocity / $c;
};
  function gamma($velocity) {
  global $c, $v;
  $b = beta($velocity);
  return 1.0 / sqrtApprox(1.0 - $b * $b);
};
  function transformation_matrix($velocity) {
  global $c, $v;
  $g = gamma($velocity);
  $b = beta($velocity);
  return [[$g, -$g * $b, 0.0, 0.0], [-$g * $b, $g, 0.0, 0.0], [0.0, 0.0, 1.0, 0.0], [0.0, 0.0, 0.0, 1.0]];
};
  function mat_vec_mul($mat, $vec) {
  global $c, $v;
  $res = [];
  $i = 0;
  while ($i < 4) {
  $row = $mat[$i];
  $value = $row[0] * $vec[0] + $row[1] * $vec[1] + $row[2] * $vec[2] + $row[3] * $vec[3];
  $res = array_merge($res, [$value]);
  $i = $i + 1;
};
  return $res;
};
  function transform($velocity, $event) {
  global $c, $v;
  $g = gamma($velocity);
  $b = beta($velocity);
  $ct = $event[0] * $c;
  $x = $event[1];
  return [$g * $ct - $g * $b * $x, -$g * $b * $ct + $g * $x, $event[2], $event[3]];
};
  echo rtrim(_str(beta($c))), PHP_EOL;
  echo rtrim(_str(beta(199792458.0))), PHP_EOL;
  echo rtrim(_str(beta(100000.0))), PHP_EOL;
  echo rtrim(_str(gamma(4.0))), PHP_EOL;
  echo rtrim(_str(gamma(100000.0))), PHP_EOL;
  echo rtrim(_str(gamma(30000000.0))), PHP_EOL;
  echo rtrim(_str(transformation_matrix(29979245.0))), PHP_EOL;
  $v = transform(29979245.0, [1.0, 2.0, 3.0, 4.0]);
  echo rtrim(_str($v)), PHP_EOL;
$__end = _now();
$__end_mem = memory_get_peak_usage(true);
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;

File: tests/algorithms/transpiler/FS/physics/lorentz_transformation_four_vector.fs
// Generated 2025-08-22 13:05 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let rec _str v =
    match box v with
    | :? float as f -> sprintf "%.10g" f
    | _ ->
        let s = sprintf "%A" v
        s.Replace("[|", "[")
         .Replace("|]", "]")
         .Replace("; ", " ")
         .Replace(";", "")
         .Replace("\"", "")
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let c: float = 299792458.0
let rec sqrtApprox (x: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable x = x
    try
        if x <= 0.0 then
            __ret <- 0.0
            raise Return
        let mutable guess: float = x / 2.0
        let mutable i: int = 0
        while i < 20 do
            guess <- (guess + (x / guess)) / 2.0
            i <- i + 1
        __ret <- guess
        raise Return
        __ret
    with
        | Return -> __ret
and beta (velocity: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable velocity = velocity
    try
        if velocity > c then
            ignore (failwith ("Speed must not exceed light speed 299,792,458 [m/s]!"))
        if velocity < 1.0 then
            ignore (failwith ("Speed must be greater than or equal to 1!"))
        __ret <- velocity / c
        raise Return
        __ret
    with
        | Return -> __ret
and gamma (velocity: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable velocity = velocity
    try
        let b: float = beta (velocity)
        __ret <- 1.0 / (sqrtApprox (1.0 - (b * b)))
        raise Return
        __ret
    with
        | Return -> __ret
and transformation_matrix (velocity: float) =
    let mutable __ret : float array array = Unchecked.defaultof<float array array>
    let mutable velocity = velocity
    try
        let g: float = gamma (velocity)
        let b: float = beta (velocity)
        __ret <- [|[|g; (-g) * b; 0.0; 0.0|]; [|(-g) * b; g; 0.0; 0.0|]; [|0.0; 0.0; 1.0; 0.0|]; [|0.0; 0.0; 0.0; 1.0|]|]
        raise Return
        __ret
    with
        | Return -> __ret
and mat_vec_mul (mat: float array array) (vec: float array) =
    let mutable __ret : float array = Unchecked.defaultof<float array>
    let mutable mat = mat
    let mutable vec = vec
    try
        let mutable res: float array = Array.empty<float>
        let mutable i: int = 0
        while i < 4 do
            let row: float array = _idx mat (int i)
            let value: float = ((((_idx row (int 0)) * (_idx vec (int 0))) + ((_idx row (int 1)) * (_idx vec (int 1)))) + ((_idx row (int 2)) * (_idx vec (int 2)))) + ((_idx row (int 3)) * (_idx vec (int 3)))
            res <- unbox<float array> (Array.append res [|value|])
            i <- i + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
and transform (velocity: float) (event: float array) =
    let mutable __ret : float array = Unchecked.defaultof<float array>
    let mutable velocity = velocity
    let mutable event = event
    try
        let g: float = gamma (velocity)
        let b: float = beta (velocity)
        let ct: float = (_idx event (int 0)) * c
        let x: float = _idx event (int 1)
        __ret <- unbox<float array> [|(g * ct) - ((g * b) * x); (((-g) * b) * ct) + (g * x); _idx event (int 2); _idx event (int 3)|]
        raise Return
        __ret
    with
        | Return -> __ret
ignore (printfn "%s" (_str (beta (c))))
ignore (printfn "%s" (_str (beta (199792458.0))))
ignore (printfn "%s" (_str (beta (100000.0))))
ignore (printfn "%s" (_str (gamma (4.0))))
ignore (printfn "%s" (_str (gamma (100000.0))))
ignore (printfn "%s" (_str (gamma (30000000.0))))
ignore (printfn "%s" (_str (transformation_matrix (29979245.0))))
let v: float array = transform (29979245.0) (unbox<float array> [|1.0; 2.0; 3.0; 4.0|])
ignore (printfn "%s" (_str (v)))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)
